File: src/main/java/yan/lx/bedrockminer/utils/BlockPlacer.java
package yan.lx.bedrockminer.utils;

import net.minecraft.block.Blocks;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.client.network.ClientPlayerInteractionManager;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemConvertible;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.network.packet.c2s.play.PlayerInteractBlockC2SPacket;
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3d;

public class BlockPlacer {
    /**
     * 简单方块放置
     *
     * @param pos  待放置位置
     * @param item 待放置方块
     */
    public static void simpleBlockPlacement(BlockPos pos, ItemConvertible item) {
        if (pos == null || item == null) {
            return;
        }

        MinecraftClient minecraftClient = MinecraftClient.getInstance();
        PlayerEntity player = minecraftClient.player;
        ClientPlayNetworkHandler clientPlayNetworkHandler = minecraftClient.getNetworkHandler();
        if (player == null || clientPlayNetworkHandler == null) {
            return;
        }

        Direction direction = Direction.UP;
        float pitch = 90f;
        minecraftClient.getNetworkHandler().sendPacket(new PlayerMoveC2SPacket.LookAndOnGround(player.getYaw(1.0f), pitch, player.isOnGround()));

        InventoryManager.switchToItem(item);
        BlockHitResult hitResult = new BlockHitResult(new Vec3d(pos.getX(), pos.getY(), pos.getZ()), direction, pos, false);

        placeBlockWithoutInteractingBlock(hitResult);
    }

    /**
     * 活塞放置
     *
     * @param pos       活塞放置坐标
     * @param direction 活塞放置方向
     */
    public static void pistonPlacement(BlockPos pos, Direction direction) {
        if (pos == null || direction == null) {
            return;
        }

        MinecraftClient minecraftClient = MinecraftClient.getInstance();
        PlayerEntity player = minecraftClient.player;
        ClientPlayNetworkHandler clientPlayNetworkHandler = minecraftClient.getNetworkHandler();

        if (player == null || clientPlayNetworkHandler == null) {
            return;
        }

        float pitch = switch (direction) {
            case UP, NORTH, SOUTH, WEST, EAST -> 90f;
            case DOWN -> -90f;
        };
        minecraftClient.getNetworkHandler().sendPacket(new PlayerMoveC2SPacket.LookAndOnGround(player.getYaw(1.0f), pitch, player.isOnGround()));

        Vec3d vec3d = new Vec3d(pos.getX(), pos.getY(), pos.getZ());

        InventoryManager.switchToItem(Blocks.PISTON);
        BlockHitResult hitResult = new BlockHitResult(vec3d, Direction.UP, pos, false);

        placeBlockWithoutInteractingBlock(hitResult);
    }

    /**
     * 放置没有交互的方块
     */
    private static void placeBlockWithoutInteractingBlock(BlockHitResult hitResult) {
        MinecraftClient minecraftClient = MinecraftClient.getInstance();
        ClientWorld world = minecraftClient.world;
        PlayerEntity player = minecraftClient.player;
        ClientPlayerInteractionManager interactionManager = minecraftClient.interactionManager;
        if (world == null || player == null || interactionManager == null) {
            return;
        }

        ItemStack itemStack = player.getStackInHand(Hand.MAIN_HAND);

        interactionManager.sendSequencedPacket(minecraftClient.world, sequence -> new PlayerInteractBlockC2SPacket(Hand.MAIN_HAND, hitResult, sequence));

        if (!itemStack.isEmpty() && !player.getItemCooldownManager().isCoolingDown(itemStack.getItem())) {
            ItemUsageContext itemUsageContext = new ItemUsageContext(player, Hand.MAIN_HAND, hitResult);
            itemStack.useOnBlock(itemUsageContext);
        }
    }
}

File: src/client/kotlin/com/github/lxyan2333/bedrockminer/client/breaking/BlockPlacer.kt
package com.github.lxyan2333.bedrockminer.client.breaking

import com.github.lxyan2333.bedrockminer.client.compat.MinecraftClientCompat
import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.InteractionHand
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3

object BlockPlacer {

    fun simpleBlockPlacement(pos: BlockPos, item: Item) {
        InteractionRangeChecker.checkRange(pos)
        val client = Minecraft.getInstance()
        val player = client.player ?: return
        val hitResult = BlockHitResult(Vec3.atBottomCenterOf(pos), Direction.DOWN, pos, false)
        //? if >= 1.21.1 {
        if (!InventoryManager.switchItemToOffHand(item)) return
        MinecraftClientCompat.useItemOn(player, InteractionHand.OFF_HAND, hitResult)
        //?} else {
        /*if (!InventoryManager.switchToItem(item)) return
        MinecraftClientCompat.useItemOn(player, InteractionHand.MAIN_HAND, hitResult)
        *///?}
    }

    fun getYawPitch(direction: Direction): Pair<Float, Float> {

        val pitch = when (direction) {
            Direction.UP -> 90f
            Direction.DOWN -> -90f
            else -> 0f
        }

        val yaw = when (direction) {
            Direction.SOUTH -> 180f
            Direction.EAST -> 90f
            Direction.NORTH -> 0f
            Direction.WEST -> -90f
            else -> 0f
        }
        return Pair(yaw, pitch)
    }

    fun vanillaPistonPlacement1(direction: Direction): Pair<Float, Float> {
        val client = Minecraft.getInstance()
        val player = client.player ?: return Pair(0.0F, 0.0F)

        val ret = getYawPitch(direction)

        client.connection?.send(MinecraftClientCompat.rotationPacket(ret.first, ret.second, MinecraftClientCompat.isOnGround(player)))

        return ret
    }

    fun vanillaPistonPlacement2(pos: BlockPos, direction: Direction) {
        InteractionRangeChecker.checkRange(pos)
        val client = Minecraft.getInstance()
        val player = client.player ?: return

        //? if >= 1.21.1 {
        if (!InventoryManager.switchItemToOffHand(Blocks.PISTON.asItem())) return
        //?} else
        //if (!InventoryManager.switchToItem(Blocks.PISTON.asItem())) return

        val hitPos = pos.relative(direction.opposite)
        val hitVec = MinecraftClientCompat.offset(Vec3.atCenterOf(hitPos), direction, 0.5)
        val hitResult = BlockHitResult(hitVec, direction, pos, false)

        // these two varialbe is the actually used value when determine the piston facing.
        // we set them on client so the piston facing is correct on client.
        val oldXRot = player.xRot
        val oldYRot = player.yRot
        val oldYHeadRot = player.yHeadRot

        val (yaw, pitch) = getYawPitch(direction)
        try {
            player.xRot = pitch
            player.yRot = yaw
            player.yHeadRot = yaw
            client.connection?.send(MinecraftClientCompat.rotationPacket(yaw, pitch, MinecraftClientCompat.isOnGround(player)))
            //? if >= 1.21.1 {
            MinecraftClientCompat.useItemOn(player, InteractionHand.OFF_HAND, hitResult)
            //?} else
            //MinecraftClientCompat.useItemOn(player, InteractionHand.MAIN_HAND, hitResult)
        } finally {
            player.xRot = oldXRot
            player.yRot = oldYRot
            player.yHeadRot = oldYHeadRot
        }
    }

    fun vanillaPistonPlacement(pos: BlockPos, direction: Direction) {
        vanillaPistonPlacement1(direction)
        vanillaPistonPlacement2(pos, direction)
    }

    /**
     * Carpet Extra accurate block placement: encodes facing direction in the
     * hit X coordinate as {@code pos.x + 2 + direction.ordinal * 2}.
     * The server-side Carpet Extra mixin decodes this to override facing.
     */
    fun carpetPistonPlacement(pos: BlockPos, direction: Direction) {
        InteractionRangeChecker.checkRange(pos)
        val client = Minecraft.getInstance()
        val player = client.player ?: return

        //? if >= 1.21.1 {
        if (!InventoryManager.switchItemToOffHand(Blocks.PISTON.asItem())) return
        //?} else
        //if (!InventoryManager.switchToItem(Blocks.PISTON.asItem())) return
        val hitVec = Vec3(pos.x + 2 + direction.ordinal * 2.0, pos.y.toDouble(), pos.z.toDouble())
        val hitResult = BlockHitResult(hitVec, direction, pos, false)

        val oldXRot = player.xRot
        val oldYRot = player.yRot
        val oldYHeadRot = player.yHeadRot
        val (yaw, pitch) = getYawPitch(direction)

        try {
            player.xRot = pitch
            player.yRot = yaw
            player.yHeadRot = yaw
            //? if >= 1.21.1 {
            MinecraftClientCompat.useItemOn(player, InteractionHand.OFF_HAND, hitResult)
            //?} else
            //MinecraftClientCompat.useItemOn(player, InteractionHand.MAIN_HAND, hitResult)
        } finally {
            player.xRot = oldXRot
            player.yRot = oldYRot
            player.yHeadRot = oldYHeadRot
        }
    }
}
